### ISSUE TITLE
Discussion: A place for code-snippets, guides and quick tips

### ISSUE THREAD

**Issue by elkowar** [xmonad/xmonad-contrib#412]
In my recent PR #403 a discussion about a general approach to sharing short bits of code like the one in said PR was started. I'll create this issue to have a dedicated place to discuss this!

## summary of what has been discussed thus far
Given the "hacky" nature of my proposed fix, it was argued that it may not be fit for inclusion in xmonad-contrib, as it is also only useful for a smaller subset of users. The initial proposal was to introduce a "Hacks" module namespace for fixes like this. 
The main proposed alternative was to use the github wiki of xmonad-contrib for things like this, which would also allow us to include gifs and images explaining any issues and features more easily.

## about using a github project wiki
As far as I know, github does not currently provide the option to do PRs to github wiki repos. Thus, using the wiki like that would have to be a rather manual process. My proposal would be to set up a separate repository which we then may be able to make github deploy to this repos project wiki via github actions

**Comment by elkowar** [comment]
Some more thoughts on this from my side:


I see a couple other issues with that approach, tho:
- having things like the fix in #403 (specifically, fixes for smaller QOL issues) in a separate place may strongly hurt discoverability when compared to having stuff be available as direct modules in contrib

- snippets and tips that are collected separate from any actual haskell project are much harder to validate, as they cannot directly be typechecked, etc. this may especially be an issue with regards to xmonad(-contrib) introducing breaking changes

- given that the whole idea of a "snippets" style repo would be that its even easier to contribute code of imperfect quality there, that could possible reduce the incentive for people to actually contribute well modularized versions of their code to xmonad-contrib. the general model of how xmonad-contrib is managed (namely, as a library) is part of the main thing I always bring up when comparing to, say, the patching approach of dwm - which a snippet-style repo would be more simmilar to.



On the other hand, having a more obvious collection of tips, explainations and snippets seems like a very good idea in general. Currently, for example, there is not really a good place where one could put explaination on how to use XMonad together with Polybar. There is a page in the Polybar wiki, but its hidden under "third party modules", and seems to be rather hard to discover. Having a official, well managed and easily contributable to place for information in a more open format would solve this quite well. We could use such a repo/wiki to share information in a blogpost-like or "quick-tip" style format in addition to sharing small fixes like the one discussed in #403. 

Having a explicitly community-contributed space for guides, tips and explainations may also encourage users to submit writeups on more general things like "how to get started with xmonad" or "how to write your first module" in a more open and informal style than is currently possible in the places where we have that kind of information currently (namely the XMonad.Docs.Contributing area, etc). This would definitely help make XMonad more approachable.

In summary, I think having a official place for writeups on tips, tricks, code snippets and explainations is a great idea, but it should not mainly be focused on simply sharing code snippets with no explaination, as that would possibly conflict with xmonad-contrib as the general place for user contributed modules and features. Also, I am of course very unsure about if something like that would actually be used much, but i cannot really judge that

**Comment by liskin** [comment]
Very well written!

I especially like your points about type checking and xmonad-contrib contribution incentives, so I'm starting to think that having XMonad.Hacks.* namespace in xmonad-contrib might be best after all. Let people submit hacks, and if anyone feels that something is more than just a hack and deserves a proper module, they can suggest that in the PR, or refactor it themselves. I guess adding gifs and images to xmonad-contrib could be doable somehow? :-)

**Comment by slotThe** [comment]
On Thu, Nov 26 2020 00:35, ElKowar wrote:
> - snippets and tips that are collected separate from any actual
> haskell project are much harder to validate, as they cannot directly
> be typechecked, etc. this may especially be an issue with regards to
> xmonad(-contrib) introducing breaking changes

This may be solved by using something like a literate Haskell file and
just adding a lot more explanations (which would also be in tune with
more of a "guide" approach).

I don't have any huge preference either way, so if you people would
rather use the XMonad.Hacks.* name space, I'd be completely fine with
that.

**Comment by elkowar** [comment]
I guess looking at `XMonad.Util.Hacks` as will be introduced by #403, we can close this for now. Having a structured collection of blog-posts or something like that may still be worth it, although the xmonad subreddit should already fill that spot, i guess.
